cylc/flow/cfgspec/suite.py

```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
+    | Fail Loudly                   | |    +-v-----------------------------+
+    +-------------------------------+ |    | * Run reverse_lookup()        |
+                                      |    | * handle reverse lookup fail  |
+                                      |    | * add platform                |
+                                      |    | * delete forbidden settings   |
+                                      |    +-------------------------------+
+                                      |
+                                      |    +-------------------------------+
+                                      +----> Return without changes        |
+                                           +-------------------------------+
+
+    Args (cfg):
+        config object to be upgraded
+
+    Returns (cfg):
+        upgraded config object
+    """
+    # If platform and old settings are set fail
+    # and remote should be added to this forbidden list
+    forbidden_with_platform = {
+        'host', 'batch system', 'batch submit command template'
+    }
+
+    for task_name, task_spec in cfg['runtime'].items():
+        # if task_name == 'delta':
+        #     breakpoint(header=f"task_name = {task_name}")
+
+        if (
+            'platform' in task_spec and 'job' in task_spec or
+            'platform' in task_spec and 'remote' in task_spec
+        ):
+            if (
+                'platform' in task_spec and
+                forbidden_with_platform & {
+                    *task_spec['job'], *task_spec['remote']
+                }
+            ):
+                # Fail Loudly and Horribly
+                LOG.error(
+                    f"A mixture of Cylc 7 (host) and Cylc 8 (platform logic)"
+                    f" should not be used. Task {task_name} set platform "
+                    f"and item in {forbidden_with_platform}"
+                )
+        elif 'platform' in task_spec:
+            # Return config unchanged
+            continue
+
+        else:
+            # Add empty dicts if appropriate sections not present.
+            if 'job' in task_spec:
+                task_spec_job = task_spec['job']
+            else:
+                task_spec_job = {}
+            if 'remote' in task_spec:
+                task_spec_remote = task_spec['remote']
+            else:
+                task_spec_remote = {}
+
+            # Deal with case where host is a function and we cannot auto
+            # upgrade at the time of loading the config.
+            if (
+                'host' in task_spec_remote and
+                REC_COMMAND.match(task_spec['remote']['host'])
+            ):
+                LOG.info(
+                    f"Unable to upgrade task '{task_name}' to platform at "
+                    f"validation because the host setting contains a "
+                    f"function. Cylc will attempt to upgrade this task on"
+                    f" job submission."
+                )
```

Comment:
For a suite with ~10'000 tasks this logging may be a bit overwhelming. We can address logging closer to release time...

Reply:
Exception instead?

Reply:
Exception instead?

Reply:
I've made this a debug statement and streamlined the text - it's now only 2 lines long!